filters/auth/grantcookie.go

```diff
@@ -41,41 +44,48 @@ func (c *cookie) isAccessTokenExpired() bool {
 	return now.After(c.Expiry)
 }
 
+// allowedForHost checks if provided host matches cookie domain
+// according to https://www.rfc-editor.org/rfc/rfc6265#section-5.1.3
+func (c *cookie) allowedForHost(host string) bool {
+	hostWithoutPort, _, err := net.SplitHostPort(host)
+	if err != nil {
+		hostWithoutPort = host
+	}
+	return strings.HasSuffix(hostWithoutPort, c.Domain)
+}
+
 // extractCookie removes and returns the OAuth Grant token cookie from a HTTP request.
 // The function supports multiple cookies with the same name and returns
 // the best match (the one that decodes properly).
 // The client may send multiple cookies if a parent domain has set a
 // cookie of the same name.
 // The grant token cookie is extracted so it does not get exposed to untrusted downstream
 // services.
-func extractCookie(request *http.Request, config OAuthConfig) (cookie *cookie, err error) {
-	old := request.Cookies()
-	new := make([]*http.Cookie, 0, len(old))
-
-	for i, c := range old {
-		if c.Name == config.TokenCookieName {
-			cookie, _ = decodeCookie(c.Value, config)
-			if cookie != nil {
-				new = append(new, old[i+1:]...)
-				break
-			}
+func extractCookie(request *http.Request, config *OAuthConfig) (*cookie, error) {
+	cookies := request.Cookies()
+	for i, c := range cookies {
+		if c.Name != config.TokenCookieName {
+			continue
 		}
-		new = append(new, c)
-	}
 
-	if cookie != nil {
-		request.Header.Del("Cookie")
-		for _, c := range new {
-			request.AddCookie(c)
+		cookie, err := decodeCookie(c.Value, config)
+		if err == nil && cookie.allowedForHost(request.Host) {
```

Comment:
Here we could also continue on err != nil
Which is maybe more obvious than the current code